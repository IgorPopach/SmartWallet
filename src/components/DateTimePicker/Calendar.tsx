import React from 'react';
import { parse } from '../../services/time';

interface Props {
    data: number;
    updateData: (value: number) => void;
}

const MONTHS = [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
];

const DAYS_SHORT = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];

const Calendar = ({ data, updateData }: Props) => {
    const parsedData = React.useMemo(() => parse(data), []);

    const [selectedDate, setSelectedDate] = React.useState(parsedData);

    const [date, setDate] = React.useState(selectedDate);

    const month = React.useMemo(() => date.month, [date]);
    const year = React.useMemo(() => date.year, [date]);

    const days = React.useMemo(() => {
        const offset = date.set({ day: 1 }).weekday;
        const daysArr = new Array(0).concat(new Array(offset - 1).fill(null));
        new Array(date.daysInMonth).fill(null).map((_, index) => daysArr.push(++index));
        return daysArr;
    }, [date]);

    const weeks = React.useMemo(
        () =>
            new Array(Math.ceil(days.length / 7)).fill(null).map((_, index) => days.slice(index * 7, (index + 1) * 7)),
        [date],
    );

    const tableBody = React.useMemo(
        () =>
            weeks.map((week, index) => (
                <tr key={index}>
                    {week.map((d, dIndex) => {
                        let styles = 'calendar-day';
                        if (d === null) {
                            return <td key={`${index} - ${dIndex}`}>{d}</td>;
                        }
                        if (d === selectedDate.day && month === selectedDate.month && year === selectedDate.year) {
                            styles = 'calendar-day is-selected';
                        }
                        const handleDayClick = () => {
                            setSelectedDate((prevState) => {
                                const updateDate = prevState.set({ day: d, month });
                                updateData(updateDate.toMillis());
                                return updateDate;
                            });
                        };
                        return (
                            <td key={`${index} - ${dIndex}`} className={styles} onClick={handleDayClick}>
                                {d}
                            </td>
                        );
                    })}
                </tr>
            )),
        [date],
    );

    const clickPrevMonth = React.useCallback(() => setDate((prevState) => prevState.minus({ month: 1 })), [date]);

    const clickNextMonth = React.useCallback(() => setDate((prevState) => prevState.plus({ month: 1 })), [date]);

    React.useEffect(() => {
        setDate(selectedDate);
    }, [selectedDate]);

    return (
        <>
            <div className="calendar-header">
                <div className="calendar-btn prev-month" onClick={clickPrevMonth}>
                    ◀
                </div>
                <div className="calendar-header-view">
                    {MONTHS[month - 1]} {year}
                </div>
                <div className="calendar-btn next-month" onClick={clickNextMonth}>
                    ▶
                </div>
            </div>
            <table className="calendar-body">
                <thead className="calendar-body__weeks">
                    <tr>
                        {DAYS_SHORT.map((d, id) => (
                            <th key={id}>{d}</th>
                        ))}
                    </tr>
                </thead>
                <tbody className="calendar-days">{tableBody}</tbody>
            </table>
        </>
    );
};

export default Calendar;
