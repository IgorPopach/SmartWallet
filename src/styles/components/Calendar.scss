.calendar {

  &-frame {
    width: 100%;
    height: 200px;
    border: 1px solid $primary;
    border-top: none;
    border-bottom-left-radius: 3px;
    border-bottom-right-radius: 3px;
  }

  &-header {
    display: flex;
    justify-content: space-around;
    padding: 3px 5px;

    &-view {
      border-bottom: 1px solid $primary;
    }
  }

  &-btn {
    &:hover {
      transition: color .3s ease;
      cursor: pointer;
      color: darken($primary, 20%);
    }
  }

  &-body {
    display: flex;
    flex-direction: column;

    &__weeks {
      padding: 3px 0;
    }

    &__days {
      border-top: 1px solid $primary;
    }

    .is-selected {
      background-color: $primary;
      color: $secondary;
    }
  }

  &-week, &-day {
    display: inline-block;
    width: 14.28%;
    text-align: center;
  }

  &-day {
    line-height: 25px;

    &:hover {
      background-color: $primary;
      color: darken($secondary, 20%);
      transition: all .3s ease;
      cursor: pointer;
    }
  }
}