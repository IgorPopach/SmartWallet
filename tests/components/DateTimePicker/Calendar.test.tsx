import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';

import Calendar from './../../../src/components/DateTimePicker/Calendar';

type Props = React.ComponentProps<typeof Calendar>;

describe('<Calendar />', () => {
    const props: Props = {
        data: 123,
        updateData: jest.fn(),
    };

    let wrapper: ShallowWrapper<Props>;

    beforeAll(() => {
        wrapper = shallow(<Calendar {...props} />);
    });

    it('Should have correct layout', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
