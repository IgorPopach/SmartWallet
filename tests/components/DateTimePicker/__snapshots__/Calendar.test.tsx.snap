// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calendar /> Should have correct layout 1`] = `
<Fragment>
  <div
    className="calendar-header"
  >
    <div
      className="calendar-btn prev-month"
      onClick={[Function]}
    >
      ◀
    </div>
    <div
      className="calendar-header-view"
    >
      January
       
      1970
    </div>
    <div
      className="calendar-btn next-month"
      onClick={[Function]}
    >
      ▶
    </div>
  </div>
  <table
    className="calendar-body"
  >
    <thead
      className="calendar-body__weeks"
    >
      <tr>
        <th
          key="0"
        >
          Sun
        </th>
        <th
          key="1"
        >
          Mon
        </th>
        <th
          key="2"
        >
          Tue
        </th>
        <th
          key="3"
        >
          Wed
        </th>
        <th
          key="4"
        >
          Thu
        </th>
        <th
          key="5"
        >
          Fri
        </th>
        <th
          key="6"
        >
          Sat
        </th>
      </tr>
    </thead>
    <tbody
      className="calendar-days"
    >
      <tr
        key="0"
      >
        <td
          key="0 - 0"
        />
        <td
          key="0 - 1"
        />
        <td
          key="0 - 2"
        />
        <td
          className="calendar-day is-selected"
          key="0 - 3"
          onClick={[Function]}
        >
          1
        </td>
        <td
          className="calendar-day"
          key="0 - 4"
          onClick={[Function]}
        >
          2
        </td>
        <td
          className="calendar-day"
          key="0 - 5"
          onClick={[Function]}
        >
          3
        </td>
        <td
          className="calendar-day"
          key="0 - 6"
          onClick={[Function]}
        >
          4
        </td>
      </tr>
      <tr
        key="1"
      >
        <td
          className="calendar-day"
          key="1 - 0"
          onClick={[Function]}
        >
          5
        </td>
        <td
          className="calendar-day"
          key="1 - 1"
          onClick={[Function]}
        >
          6
        </td>
        <td
          className="calendar-day"
          key="1 - 2"
          onClick={[Function]}
        >
          7
        </td>
        <td
          className="calendar-day"
          key="1 - 3"
          onClick={[Function]}
        >
          8
        </td>
        <td
          className="calendar-day"
          key="1 - 4"
          onClick={[Function]}
        >
          9
        </td>
        <td
          className="calendar-day"
          key="1 - 5"
          onClick={[Function]}
        >
          10
        </td>
        <td
          className="calendar-day"
          key="1 - 6"
          onClick={[Function]}
        >
          11
        </td>
      </tr>
      <tr
        key="2"
      >
        <td
          className="calendar-day"
          key="2 - 0"
          onClick={[Function]}
        >
          12
        </td>
        <td
          className="calendar-day"
          key="2 - 1"
          onClick={[Function]}
        >
          13
        </td>
        <td
          className="calendar-day"
          key="2 - 2"
          onClick={[Function]}
        >
          14
        </td>
        <td
          className="calendar-day"
          key="2 - 3"
          onClick={[Function]}
        >
          15
        </td>
        <td
          className="calendar-day"
          key="2 - 4"
          onClick={[Function]}
        >
          16
        </td>
        <td
          className="calendar-day"
          key="2 - 5"
          onClick={[Function]}
        >
          17
        </td>
        <td
          className="calendar-day"
          key="2 - 6"
          onClick={[Function]}
        >
          18
        </td>
      </tr>
      <tr
        key="3"
      >
        <td
          className="calendar-day"
          key="3 - 0"
          onClick={[Function]}
        >
          19
        </td>
        <td
          className="calendar-day"
          key="3 - 1"
          onClick={[Function]}
        >
          20
        </td>
        <td
          className="calendar-day"
          key="3 - 2"
          onClick={[Function]}
        >
          21
        </td>
        <td
          className="calendar-day"
          key="3 - 3"
          onClick={[Function]}
        >
          22
        </td>
        <td
          className="calendar-day"
          key="3 - 4"
          onClick={[Function]}
        >
          23
        </td>
        <td
          className="calendar-day"
          key="3 - 5"
          onClick={[Function]}
        >
          24
        </td>
        <td
          className="calendar-day"
          key="3 - 6"
          onClick={[Function]}
        >
          25
        </td>
      </tr>
      <tr
        key="4"
      >
        <td
          className="calendar-day"
          key="4 - 0"
          onClick={[Function]}
        >
          26
        </td>
        <td
          className="calendar-day"
          key="4 - 1"
          onClick={[Function]}
        >
          27
        </td>
        <td
          className="calendar-day"
          key="4 - 2"
          onClick={[Function]}
        >
          28
        </td>
        <td
          className="calendar-day"
          key="4 - 3"
          onClick={[Function]}
        >
          29
        </td>
        <td
          className="calendar-day"
          key="4 - 4"
          onClick={[Function]}
        >
          30
        </td>
        <td
          className="calendar-day"
          key="4 - 5"
          onClick={[Function]}
        >
          31
        </td>
      </tr>
    </tbody>
  </table>
</Fragment>
`;
